### ISSUE TITLE
proposal for streamlining creation of dds objects

### ISSUE THREAD

**Issue by daler** [lcdb/lcdb-wf#275]
We typically make a handful of dds objects to look at different aspects of the experiment. This ends up creating a fair amount of code duplication, with the corresponding introduction of possible errors.

For example, it's currently pretty messy to remove outliers, especially in cases where the technical replicates need to be merged and ensembl versions need to be stripped. That is, these three things happen:

```r
#1.
dds <- DESeqDataSetFromCombinedFeatureCounts(
  '../data/rnaseq_aggregation/featurecounts.txt',
  sampletable=colData,
  design=~group)

#2.
dds <- collapseReplicates(dds, dds$biorep)

#3.
if (strip.dotted.version){
  rownames(dds) <- sapply(strsplit(rownames(dds), '.', fixed=TRUE),
                                  function (x) x[1])
}
```                                        

When creating different dds objects, the above three steps need to happen each time and it's easy to get mixed up with the variable names. I think one way to streamline this would be to have DESeqDataSetFromCombinedFeatureCounts do the collapsing (if needed) and version strip (if needed) within that function.

The result of this example code is a list of dds objects where they all pull from the same salmon input files and use the same technical replicate collapsing and the same Ensembl version removal.

```r
lst <- list(
  main=list(colData, ~group),
  no.rep.2=list(colData %>% filter(replicate!=2), ~group),
  interaction=list(colData %>% filter(replicate!=2), ~genotype + time + genotype:time)
)

dds.list <- make.dds.list(
   lst, salmon.files=salmon.files, combine.by='biorep', remove.version=TRUE)
```

In this way, `make.dds.list` only needs to know the salmon files, collapse info, and version removal info once. The only things that change are the colData and model.

So . . . just have to write `make.dds.list` to make that happen!

An additional side effect is that the res.list object will no longer store copies of each dds object, we can just refer by list name to the new dds.list object.

Note: this will use the salmon importing through tximport, which has arguments for at least removing ensembl versions, so that can be passed through to the tximport function.

**Comment by njohnso6** [comment]
Benchmarking, if done correctly, revelas lscratch, normal, and shm to have similar times:
Normal? Scratch allocated but not assigned:
user        sys        elapsed
55.523    3.711    62.643
56.57    1.37        57.956

Lscrath officially allocated 
user        sys        elapsed   
59.886    3.770    64.042
52.935    1.527    54.465

With shm
57.783    3.90    62.232
50.792    1.278    52.076

**Comment by njohnso6** [comment]
With this, I want to test out using ... If I pass ... will every function within a function get those same arguments? Will that function get mad if some of those arguments are irrelevant to it? What if the arguments have the same keywords?

**Comment by daler** [comment]
You can convert `...` to a list inside the function, see [here](https://www.burns-stat.com/the-three-dots-construct-in-r/) for some more info. That also references [Circle 8.3.15 of the R Inferno](https://www.burns-stat.com/pages/Tutor/R_inferno.pdf) some more ideas, but probably the best way to figure out the particular behavior for the case you have in mind is to test it!

**Comment by njohnso6** [comment]
This has the basic workings within the issue-275 branch. However, nothing is set up to utilize the dds.list yet. That is the next step.

**Comment by njohnso6** [comment]
This particular issue is now complete. A subsequent issue is suggested to deal allow salmon input with subset.counts.

**Comment by daler** [comment]
Closed in #291 (thanks Nick!)
